File: Stacks/Valid Parentheses/ValidParentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> check = new Stack<Character>();

        for(int i = 0; i < s.length(); i++){
            if( (s.charAt(i) == '(') || (s.charAt(i) == '[') || (s.charAt(i) == '{') ){
                check.push(s.charAt(i));
            }
            else{
                switch(s.charAt(i)){
                    case ')':
                        if(check.empty() || check.pop() != '('){
                            return false;
                        }
                        break;
                    case ']':
                        if(check.empty() || check.pop() != '['){
                            return false;
                        }
                        break;
                    case '}':
                        if(check.empty() || check.pop() != '{'){
                            return false;
                        }
                        break;
                    default: 
                        break;
                }
            }
        }

        return check.empty();
    }
}


// {][()([])]}
//{)([]}        {()[]}
//    |       |
//    |       |
//    |       |   
//    |       |   
//    |       | 
//  } |       | {

File: Stacks/Valid Parentheses/validParentheses.py
from collections import deque

class Solution:
    def isValid(self, s: str) -> bool:
        tempStack = deque()

        for char in s:
            if char in '([{':
                tempStack.append(char)
            elif char == ')': 
                if not tempStack or tempStack[-1] != '(':
                    return False
                tempStack.pop()
            elif char == ']':  
                if not tempStack or tempStack[-1] != '[':
                    return False
                tempStack.pop()
            elif char == '}': 
                if not tempStack or tempStack[-1] != '{':
                    return False
                tempStack.pop()
    
        if not tempStack:
            return True
        else:
            return False